Documentation/performance/JitOptimizerTodoAssessment.md

```diff
@@ -0,0 +1,132 @@
+difficult to avoid at source level.  The recent work on finally cloning, empty
+finally removal, and empty try removal targeted this.  Writethrough is another
+key optimization here, and we are actively pursuing it.  Other things we've
+discussed include inlining methods with EH and computing funclet callee-save
+register usage independently of main function callee-save register usage, but
+I don't think we have any particular data pointing to either as a high priority.
+
+
+Loop Optimizations
+------------------
+
+We haven't been targeting benchmarks that spend a lot of time doing compuations
+in an inner loop.  Pursuing loop optimizations for the peanut butter effect
+would seem odd.  So this simply hasn't bubbled up in priority yet, though it's
+bound to eventually.
+
+
+More Expression Optimizations
+-----------------------------
+
+We again don't have particular benchmarks pointing to key missing cases, and
+balancing the CQ vs TP will be delicate here, so it would really help to have
+an appropriate benchmark suite to evaluate this work against.
+
+
+Forward Substitution
+--------------------
+
+This too needs an appropriate benchmark suite that I don't think we have at
+this time.  The tradeoffs against register pressure increase and throughput
+need to be evaluated.  This also might make more sense to do if/when we can
+handle SSA renames.
+
+
+Value Number Conservativism
+---------------------------
+
+We have some frustrating phase-ordering issues resulting from this, but the
+opt-repeat experiment indicated that they're not prevalent enough to merit
+pursuing changing this right now.  Also, using SSA def as the proxy for value
+number would require handling SSA renaming, so there's a big dependency chained
+to this.
+Maybe it's worth reconsidering the priority based on throughput?
+
+
+High Tier Optimizations
+-----------------------
+
+We don't have that many knobs we can "crank up" (though we do have the tracked
+assertion count and could switch inliner policies), nor do we have any sort of
+benchmarking story set up to validate whether tiering changes are helping or
+hurting.  We should get that benchmarking story sorted out and at least hook
+up those two knobs.
+
+
+Low Tier Back-Off
+-----------------
+
+We have some changes we know we want to make here: morph does more than it needs
+to in minopts, and tier 0 should be doing throughput-improving inlines, as
+opposed to minopts which does no inlining.  It would be nice to have the
+benchmarking story set up to measure the effect of such changes when they go in,
+we should do that.
+
+
+Async
+-----
+
+We've made note of the prevalence of async/await in modern code (and particularly
+in web server code such as TechEmpower), but haven't really identified concrete
+opportunities presented here, to my knowledge.  Some sort of study of async
+peanut butter is probably in order, but what would that look like?
```

Comment:
>  At that point its going async, then memory vs register becomes a minimal factor. 

Yes, that's the idea, to move some loads/stores on the "going async" path. Though it looks like Roslyn is in a better position to do this. For example:
```C#
static async Task Test()
{
    int i;
    for (i = 0; i < 10; i++)
        Whatever(i);
    await Task.CompletedTask;
    for (; i < 20; i++)
        Whatever(i);
}
```
In this example `i` becomes a struct field and every access to it turns into a memory access. To make things worse all this is in a try/catch block. Accessing `i` requires `this` and this is live in the catch block and this cannot be enregistered. So it's actually 2 memory accesses for every access to `i`:
```asm
G_M33543_IG04:
       488B4D10             mov      rcx, bword ptr [rbp+10H]
       8B4904               mov      ecx, dword ptr [rcx+4]
       E82AFAFFFF           call     Program:Whatever(int)
       488B4D10             mov      rcx, bword ptr [rbp+10H]
       8B7104               mov      esi, dword ptr [rcx+4]
       FFC6                 inc      esi
       488B4D10             mov      rcx, bword ptr [rbp+10H]
       897104               mov      dword ptr [rcx+4], esi
       488B4D10             mov      rcx, bword ptr [rbp+10H]
       8379040A             cmp      dword ptr [rcx+4], 10
       7CDA                 jl       SHORT G_M33543_IG04
```
Perhaps Roslyn could generate something like:
```C#
static async Task Test()
{
    int i;
    for (i = 0; i < 10; i++)
        Whatever(i);
    this.i = i; // "spill" i to a struct field
    await Task.CompletedTask;
    i = this.i; // "unspill" i from the struct field
    for (; i < 20; i++)
        Whatever(i);
}
```